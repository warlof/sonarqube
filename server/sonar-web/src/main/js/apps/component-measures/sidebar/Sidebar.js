/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import React from 'react';
import DomainFacet from './DomainFacet';
import { groupByDomains } from '../utils';
import type { MeasureEnhanced } from '../../../components/measure/types';
import type { Query } from '../types';

type Props = {|
  measures: Array<MeasureEnhanced>,
  selectedMetric: string,
  updateQuery: Query => void
|};

type State = {|
  closedFacets: { [string]: boolean },
  measuresByDomains: Array<{ name: string, measures: Array<MeasureEnhanced> }>
|};

export default class Sidebar extends React.PureComponent {
  props: Props;
  state: State;

  constructor(props: Props) {
    super(props);
    this.state = {
      closedFacets: {},
      measuresByDomains: groupByDomains(props.measures)
    };
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.measures !== this.props.measures) {
      this.setState({ measuresByDomains: groupByDomains(nextProps.measures) });
    }
  }

  toggleFacet = (name: string) => {
    this.setState(({ closedFacets }) => ({
      closedFacets: { ...closedFacets, [name]: !closedFacets[name] }
    }));
  };

  changeMetric = (metric: string) => this.props.updateQuery({ metric });

  render() {
    const { measuresByDomains } = this.state;
    return (
      <div className="search-navigator-facets-list">
        {measuresByDomains.map(domain =>
          <DomainFacet
            key={domain.name}
            domain={domain}
            onChange={this.changeMetric}
            onToggle={this.toggleFacet}
            open={!this.state.closedFacets[domain.name]}
            selected={this.props.selectedMetric}
          />
        )}
      </div>
    );
  }
}
