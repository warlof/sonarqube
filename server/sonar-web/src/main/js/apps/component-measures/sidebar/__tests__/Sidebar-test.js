/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import { shallow } from 'enzyme';
import Sidebar from '../Sidebar';

const MEASURES = [
  {
    metric: {
      key: 'lines_to_cover',
      type: 'INT',
      name: 'Lines to Cover',
      domain: 'Coverage'
    },
    value: '431',
    periods: [{ index: 1, value: '70' }],
    leak: '70'
  },
  {
    metric: {
      key: 'coverage',
      type: 'PERCENT',
      name: 'Coverage',
      domain: 'Coverage'
    },
    value: '99.3',
    periods: [{ index: 1, value: '0.0999999999999943' }],
    leak: '0.0999999999999943'
  },
  {
    metric: {
      key: 'duplicated_lines_density',
      type: 'PERCENT',
      name: 'Duplicated Lines (%)',
      domain: 'Duplications'
    },
    value: '3.2',
    periods: [{ index: 1, value: '0.0' }],
    leak: '0.0'
  }
];

const PROPS = {
  measures: MEASURES,
  selectedMetric: 'foo',
  updateQuery: () => {}
};

it('should display two facets', () => {
  expect(shallow(<Sidebar {...PROPS} />)).toMatchSnapshot();
});

it('should correctly toggle facets', () => {
  const wrapper = shallow(<Sidebar {...PROPS} />);
  expect(wrapper.state('closedFacets').bugs).toBeUndefined();
  wrapper.instance().toggleFacet('bugs');
  expect(wrapper.state('closedFacets').bugs).toBeTruthy();
  wrapper.instance().toggleFacet('bugs');
  expect(wrapper.state('closedFacets').bugs).toBeFalsy();
});
