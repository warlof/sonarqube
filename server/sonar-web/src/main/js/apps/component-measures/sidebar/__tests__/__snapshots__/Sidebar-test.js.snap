// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display two facets 1`] = `
<div
  className="search-navigator-facets-list"
>
  <DomainFacet
    domain={
      Object {
        "measures": Array [
          Object {
            "leak": "0.0999999999999943",
            "metric": Object {
              "domain": "Coverage",
              "key": "coverage",
              "name": "Coverage",
              "type": "PERCENT",
            },
            "periods": Array [
              Object {
                "index": 1,
                "value": "0.0999999999999943",
              },
            ],
            "value": "99.3",
          },
          Object {
            "leak": "70",
            "metric": Object {
              "domain": "Coverage",
              "key": "lines_to_cover",
              "name": "Lines to Cover",
              "type": "INT",
            },
            "periods": Array [
              Object {
                "index": 1,
                "value": "70",
              },
            ],
            "value": "431",
          },
        ],
        "name": "Coverage",
      }
    }
    onChange={[Function]}
    onToggle={[Function]}
    open={true}
    selected="foo"
  />
  <DomainFacet
    domain={
      Object {
        "measures": Array [
          Object {
            "leak": "0.0",
            "metric": Object {
              "domain": "Duplications",
              "key": "duplicated_lines_density",
              "name": "Duplicated Lines (%)",
              "type": "PERCENT",
            },
            "periods": Array [
              Object {
                "index": 1,
                "value": "0.0",
              },
            ],
            "value": "3.2",
          },
        ],
        "name": "Duplications",
      }
    }
    onChange={[Function]}
    onToggle={[Function]}
    open={true}
    selected="foo"
  />
</div>
`;
